server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -282,10 +317,44 @@
             }
             throw infe;
         }
+        checkSystemIndexAccess(context, metadata, concreteIndices);
         return concreteIndices.toArray(new Index[concreteIndices.size()]);
     }
 
+    private static void checkSystemIndexAccess(Context context, Metadata metadata, Set<Index> concreteIndices) {
+        if (context.isSystemIndexAccessAllowed() == false) {
+            final List<String> resolvedSystemIndices = concreteIndices.stream()
+                .map(metadata::index)
+                .filter(IndexMetadata::isSystem)
+                .map(i -> i.getIndex().getName())
+                .sorted() // reliable order for testing
+                .collect(Collectors.toList());
+            if (resolvedSystemIndices.isEmpty() == false) {
+                switch (SYSTEM_INDEX_ACCESS_BEHAVIOR) {
+                    case ALLOWED:
+                        // If access is allowed in all cases, then there's nothing to do here.
+                        break;
+                    case DEPRECATED:
+                        deprecationLogger.deprecate("open_system_index_access",
+                            "this request accesses system indices: {}, but in a future major version, direct access to system " +
+                                "indices will be prevented by default", resolvedSystemIndices);
+                        break;
+                    case RESTRICTED:
+                        final String errorMsg = "while resolving indices, resolved system indices [" + resolvedSystemIndices +
+                            "] when access to system indices is restricted";
+                        assert false : errorMsg;
```

Comment:
Should this be removed?

Reply:
This is intentional. This will cause the node to fail if this line is hit with assertions enabled (i.e. during testing), which is what I want as this would indicate there's a bug in the code which should cause system indices to be completely omitted - and if there is, I want it to yell as loudly as possible.

If this line is hit without assertions enabled, it will fall through to the next line, which throws an exception. This is a relatively common pattern to enforce invariants more strictly during testing than at runtime.